1. Pick one of the datasets from the ChatBot session(s) of the TUT demo (or from your own ChatBot session if you wish) and use the code produced through the ChatBot interactions to import the data and confirm that the dataset has missing values

In [1]:
# feel free to just use the following if you prefer...
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [2]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

2. Start a new ChatBot session with an initial prompt introducing the dataset you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then

2.1 use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,

2.2 write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot to explain these terms in the context of your dataset

In [3]:
df.columns

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

In [4]:
df.shape #(2.1)

(391, 11)

There are 391 rows, 11 columns

2.2- 

Observations:  
      Observations are specific sets of data that correspond to each individual entry in the entire dataset. Each row of data is an observation, and each observation is all of the data that makes up that row.
      
Variables:
      Variables correspond to the columns in a dataset and each contain a different piece of data for each observation. 

3. Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to provide these summaries for your dataset

In [5]:
categorical_columns = df.select_dtypes(include=['object']).columns
for col in categorical_columns:
    print(df[col].value_counts())

id
admiral    1
mott       1
paula      1
patty      1
pate       1
          ..
eloise     1
elmer      1
ellie      1
elise      1
zucker     1
Name: count, Length: 390, dtype: int64
name
Admiral    1
Muffy      1
Paula      1
Patty      1
Pate       1
          ..
Elvis      1
Eloise     1
Elmer      1
Ellie      1
Zucker     1
Name: count, Length: 391, dtype: int64
gender
male      204
female    187
Name: count, dtype: int64
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3

In [10]:
numeric_columns = df.select_dtypes(include=['number']).columns
print(df[numeric_columns].describe())

            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000


In [9]:
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

only numeric column is row_n, so that is not a good summary. also the summary of the categorical data is not super helpful since there is very little repition. Good to know how to do this for other data though.

4. If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain (perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in the "count" column

In [12]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [13]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [14]:
df.shape

(891, 15)

In [15]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

there are 891 rows, but only 714 results return in the 'count' row for the 'age' column. This is a difference of 177. Because we already ran "df.isna().sum()" we know that there are 177 missing values in the age column. As a result, the "df.describe()" command only returns the count of the 714 values that do exist in the dataset.

(double checked w chatGPT, it says that assessment is correct)

5. Use your ChatBot session to help understand the difference between the following and then provide your own paraphrasing summarization of that difference

- an "attribute", such as df.shape which does not end with ()
- and a "method", such as df.describe() which does end with ()

An attribute is a value of the dataset that are not associated with functions. df.shape is a good example of this because it is simply a detail of the dataset, rather than a function being performed to return information about the dataset.

A method is a function that returns information about the dataset. They are associated with actions or processes, which is why they need to have parentheses, even if they are not being passed an argument.

6. The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

   'count': the count of values entered for this variable in the dataset (amt of observations which have a value for this variable)
   'mean': the average of the data for this variable
   'std': the standard deviation of the sample, aka the average distance from the mean for this variable
   'min': the minimum value out of all the values for this variable
   '25%': the first quartile, also known as the 25th percentile. a quarter of values are lower than it
   '50%': the second quartile, also known as the median. half of values are lower than it
   '75%': the third quartile, also known as the 75th percentile. a quarter of values are higher than it
   'max': the maximum value out of all values for this variable

7. Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

7.1 Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']

   If most columns are complete, but one or two observations are missing most data, using df.dropna() would allow you to just remove the problematic observations, and leave as much data intact as possible

7.2 Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()

   If one specific variable was missing data for many of the observations, it would allow that column to be removed, and none of the observations would lose their other data. However, this comes at the loss of a whole variable.

7.3 Discuss why applying del df['col'] before df.dropna() when both are used together could be important

   If one variable was missing data for a large amount of the observations, it would make it so that any observations who were missing only that variable were kept intact when df.dropna() was used. Otherwise, all the observations missing only that specific variable would be deleted by df.dropna()

7.4 Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [4]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

7.4: Because deck is empty for so many of them, we are going to remove that specific column. Because embark_town and embarked are empty for only two, we can have those removed with df.dropna()
The real question is do we want to lose 177 observations because they lack age, or do we want to scrap age altogether. I am deciding to get rid of age, and keep as many observations as we can for more concrete data.

In [12]:
df = pd.read_csv(url)
del df['deck']
del df['age']
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

In [17]:
df.dropna()

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male  embark_town alive  alone  
0          True  Southampton    no  False  
1         False    Cherbourg   yes  False  
2         False  Southampton   yes   True  
3         False  Southampton   yes  False  
4          True  Southampton    no   True  
..          ...          ...   ...    ...  
886        True  Southampton    no   True  
887       False  Southampton   yes   True  
888       False  Southampton    no  False  
889        True    Cherbourg   yes   True  
890        True   Queenstown    no   True  

[889 rows x 13 columns]

In [21]:
df = df.dropna()  

In [24]:
df.isna().sum()

survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [25]:
df.shape

(889, 13)

   7.4 contd. As can see, the data has now been cleared up to the point where there are no missing values! This comes at the cost of only 2 rows and 2 columns of data. 

8. Give brief explanations in your own words for any requested answers to the questions below
  
  8.1 Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you
  
This funtion essentially takes all of the data in "col1" and groups them by value (especially helpful if col1 is categorical) and then describes "col2" for each of these groups. It helps to provide data on the relationship between these two variables in a dataset. An example below is an analysis of the data that compares the town that the passenger embarked from with the fare paid.

  8.2 Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?
  
Ultimately the missing values in the first context, df.describe, represent the total missing values for that column, whereas when grouping by column, we can see which groups have the most non-missing counts. In fact, seeing the amount of surving data that exist for each group can help give us insight into that variable.

In [3]:
# example analysis for 8.1
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
del df['deck']
del df['age']
df = df.dropna()  
df.groupby("embark_town")["fare"].describe()

count       mean        std     min       25%    50%        75%  \
embark_town                                                                    
Cherbourg    168.0  59.954144  83.912994  4.0125  13.69795  29.70  78.500025   
Queenstown    77.0  13.276030  14.188047  6.7500   7.75000   7.75  15.500000   
Southampton  644.0  27.079812  35.887993  0.0000   8.05000  13.00  27.900000   

                  max  
embark_town            
Cherbourg    512.3292  
Queenstown    90.0000  
Southampton  263.0000

  8.3 Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT
  
  (assessment below)

In [1]:
#A. Forget to include import pandas as pd in your code

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df

#I found for this one that ChatGPT was more helpful in quickly and concisely explaining the problem
# however it only would have taken a couple seconds of scrolling on google to find a similar answer

NameError: name 'pd' is not defined

In [2]:
#B Mistype "titanic.csv" as "titanics.csv"

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)
df

#for this one ChatGPT was very helpful in giving me the right answer, but google was not very helpful at all.

HTTPError: HTTP Error 404: Not Found

In [3]:
#C Try to use a dataframe before it's been assigned into the variable

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
DF.groupby("col1")["col2"].describe()

# this was more interesting as a google search gave me a much more broad definition of the error and how to fix it
# while chatGPT gave me a specific one. However, google was much easier to digest, which is something to consider

NameError: name 'DF' is not defined

In [4]:
#D Forget one of the parentheses somewhere the code

df = pd.read_csv(url
                 
# Google gives a couple different answers from a few sites, ChatGPT gives the same info a bit faster and
# easier to process. Ultimately ChatGPT is better in this case

SyntaxError: incomplete input (3688783506.py, line 3)

In [5]:
#E Mistype one of the names of the chained functions with the code

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.groupby("col1")["col2"].describle()

#ChatGPT did a great job of explaining what it looked like was wrong with my code and how to fix it. Google was not
# as helpful and once again, is less easy to digest and also less specific to my problem.

KeyError: 'col1'

In [7]:
#F Use a column name that's not in your data for the groupby and column selection

import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.groupby("Sex")["Age"].describe()

#Just like the last one, ChatGPT gives a specific and correct explanation of the problem and how to fix it. Google
# because it can not read the code and adapt to it, does not give such specificity.

KeyError: 'Sex'

In [8]:
#G Forget to put the column name as a string in quotes for the groupby and column selection, and see if the 
# ChatBot and google are still as helpful as they were for the previous question

df.groupby(sex)["age"].describe()

#here google actually was easy to find an answer quickly, but at the same time chatGPT gives a specific code
# to replace mine with and an explanation as to why

NameError: name 'sex' is not defined

  8.3 Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT
  
  
I ultimately think that while Google can be helpful in some contexts (say to find the definition of an error without context), I think that ChatGPT proves much more helpful because it provides insight as to why the code you have provided is faulty, explains why the code creates an error in the computer, and gives code that can replace the faulty code. As for speed, google can obvsously execute a query faster than ChatGPT can, but given the time it would take to sift through a couple results, it is clear that ChatGPT is faster and more efficient for getting answers. As a result, I am tempted to want to use ChatGPT more than google to help me troubleshoot problems with my code, as it will not only specifically explain to me where I have gone wrong, but help me fix it.

9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Yes!

CHATBOT SUMMARY 1
Summary of Interactions
1. Dataset Overview:

Initial Inquiry: You asked about how to explore a dataset downloaded from a URL, specifically to understand its columns and the amount of data it contains.
Solution: I provided Python code using pandas to load the dataset and inspect its columns and dimensions.
2. Observations and Variables:

Observations: Explained that an observation in the dataset refers to a single row, representing one entity or data point.
Variables: Explained that variables are columns in the dataset, representing different attributes or characteristics of the observations.
3. Summarizing Columns:

Attributes vs. Methods:
Attributes: Properties of an object that do not require parentheses to access (e.g., df.shape).
Methods: Functions associated with an object that require parentheses to call (e.g., df.describe()).
Summary Statistics: Provided methods to get a summary of columns, including data types, basic statistics, and unique values.
4. Handling NaN Values:

df.dropna() vs. del df['col']:
df.dropna(): Removes rows or columns with NaN values. Parameters can be adjusted to specify which rows or columns to drop.
del df['col']: Removes a specific column from the DataFrame.
Issue with Persistent NaN Values: Discussed why NaN values might still appear after using df.dropna() and provided solutions, including assigning the result back to the DataFrame or using inplace=True.
5. Making Changes Permanent:

Assignment: Explained that assigning the result of df.dropna() back to the DataFrame (e.g., df = df.dropna()) makes the change permanent.
In-Place Modification: Using inplace=True (e.g., df.dropna(inplace=True)) directly modifies the original DataFrame.

https://chatgpt.com/share/40b44fdf-14b4-4706-8264-1b8ba0af49a7

CHATBOT SUMMARY 2

Summary:
Initial Inquiry:
You asked for help with initial analyses of the Titanic dataset, specifically understanding the code and analysis you're performing.
df.groupby("col1")["col2"].describe() Explanation:
I explained that df.groupby("col1")["col2"].describe() groups the DataFrame by "col1" and then provides descriptive statistics for "col2" within each group.
Difference in df.describe() vs. df.groupby("col1")["col2"].describe():
I outlined how df.describe() provides statistics for each column in the entire DataFrame, while df.groupby("col1")["col2"].describe() provides statistics for "col2" within each group defined by "col1", which captures different aspects of the data.
Handling NameError: name 'pd' is not defined:
I suggested importing pandas with import pandas as pd and ensured that pandas is installed.
Handling HTTPError: HTTP Error 404: Not Found:
I explained that this error occurs when the URL is incorrect or the resource is not found, and provided ways to handle it programmatically.
Handling SyntaxError: incomplete input:
I detailed common causes of this error, such as unclosed parentheses, incomplete statements, and missing function bodies, and provided examples of how to fix them.
Handling KeyError:
I addressed how KeyError indicates that a column name is not found in the DataFrame, and demonstrated how to use the correct column names, as well as fixing the typo describle() to describe().
Handling NameError: name 'DF' is not defined:
I explained that this error occurs due to case sensitivity in variable names and suggested using the correct case (e.g., df instead of DF).
Handling NameError with column names as variables:
I clarified that column names should be passed as strings, not as variables, and provided the corrected code for grouping and describing data.
Key Points:
Ensure correct column names and casing in your DataFrame.
Use strings for column names in methods like groupby.
Handle missing imports, HTTP errors, and syntax issues by following best practices and checking for common mistakes.

https://chatgpt.com/share/585dda45-77fe-4faa-9e80-1f57f5f87261